### PR TITLE
Puppi pileupjetID

### PR DESCRIPTION
**Pull request by knash**
Update to pileupjetID algorithm to include an option to calculate the discriminator variables using Puppi weights.

There are a few effects, but one visible change is the fractional pt variables. previously these were calculated using the Puppi jet pt in the denominator and and the unweighted constituent sum in the numerator which leads to an undesirable >1 tail in the total pt fraction. See below

OFF

image

ON

image

There was no additional implementation beyond the feature addition (ie no standard workflows should change). 

I think the changes in RecoJets/JetProducers/src/PileupJetIdAlgo.cc, Line 509-510 are appropriate but would like a second opinion @ahinzmann .

This was tested only using a modified version of the script:
RecoJets/JetProducers/test/testJetTools_onMiniAOD_cfg.py
There were a few small changes required for this script to run in 10_3_X which I am also pushing.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ahinzmann** [comment]
@knash l905-910 work with the current official MiniAOD, but maybe not custom made MiniAODs. I'd suggest, if usePuppi is activated, to recompute neutral multiplicity by hand.

File: RecoJets/JetProducers/src/PileupJetIdAlgo.cc [hunk 1]
```diff
@@ -367,7 +370,10 @@ PileupJetIdentifier PileupJetIdAlgo::computeIdVariables(const reco::Jet * jet, f
 			if( icone < ncones ) { *coneEmFracs[icone] += candPt; }
 			internalId_.ptDNe_    += candPt*candPt;
 			sumPtNe               += candPt;
+			multNeut += candPuppiWeight;
 		}
+		if((abs(icand->pdgId()) == 1) || (abs(icand->pdgId()) == 2)) multNeut += candPuppiWeight;
```
Comment (ahinzmann) [inline RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: What is the purpose of this line? PdgId 1,2 seem not to be used in PF? https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideParticleFlow#Particle_types

Comment (knash) [reply RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: These are HF had/gamma. The multNeut should reproduce patjet->neutralMultiplicity() exactly in the case that usePuppi is False.

So we have events like this (setting usePuppi to False)

Adding 130
Adding 130
Adding 22
Adding 22
Adding 22
Adding 22
Adding 22
Adding 22
multNeut = 8  ; patjet->neutralMultiplicity() = 8

and events like this 

Adding 1
Adding 1
Adding 1
Adding 1
Adding 1
multNeut = 5 ;  patjet->neutralMultiplicity() = 5

Comment (ahinzmann) [reply RecoJets/JetProducers/src/PileupJetIdAlgo.cc]: ok makes sense!

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24287	e5fdda6
- code changes are in line with the PR description; current defaults are unchanged
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by fabiocos** [comment]
+1

**Pull request merged**